# Applying Advanced Transformations (Core)

## The Data
You will be working with a heavily modified version of the Superheroes dataset from Kaggle.

The dataset includes two csv's:

- superhero_info.csv:
Contains Name, Publisher, Demographic Info, and Body measurements.
- superhero_powers.csv:
Contains Hero name and list of powers

## The Task
Your task is two-fold:

I. Clean the files and combine them into one final DataFrame.

This dataframe should have the following columns:
- Hero (Just the name of the Hero)
- Publisher
- Gender
- Eye color
- Race
- Hair color
- Height (numeric)
- Skin color
- Alignment
- Weight (numeric)
- Plus, one-hot-encoded columns for every power that appears in the dataset. E.g.:
    - Agility
    - Flight
    - Superspeed
    - etc.

Hint: There is a space in "100 kg" or "52.5 cm"

II. Use your combined DataFrame to answer the following questions.

1. Compare the average weight of super powers who have Super Speed to those who do not.
2. What is the average height of heroes for each publisher?

## Imports

In [1]:
import pandas as pd
import numpy as np
import os, json

### Import and clean powers csv 

In [2]:
df_powers = pd.read_csv('Data/superhero_powers - superhero_powers.csv')
df_powers.info()
df_powers.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [3]:
df_powers['Powers'].value_counts()

Intelligence                                                                                                                                                                                                                                                         8
Durability,Super Strength                                                                                                                                                                                                                                            5
Agility,Stealth,Marksmanship,Weapons Master,Stamina                                                                                                                                                                                                                  4
Marksmanship                                                                                                                                                                                                       

In [4]:
powers = df_powers['Powers'].str.split(",")

In [5]:
powers.head()

0      [Agility, Super Strength, Stamina, Super Speed]
1    [Accelerated Healing, Durability, Longevity, S...
2    [Agility, Accelerated Healing, Cold Resistance...
3                                 [Lantern Power Ring]
4    [Accelerated Healing, Intelligence, Super Stre...
Name: Powers, dtype: object

In [6]:
exploded = powers.explode("Powers")
exploded.head()

0                Agility
1         Super Strength
2                Stamina
3            Super Speed
4    Accelerated Healing
Name: Powers, dtype: object

In [7]:
## saving the unique values from the exploded column
cols_to_make = exploded.dropna().unique()
cols_to_make.sort()
cols_to_make

array(['Accelerated Healing', 'Adaptation', 'Agility',
       'Animal Attributes', 'Animal Control', 'Animal Oriented Powers',
       'Animation', 'Anti-Gravity', 'Astral Projection', 'Astral Travel',
       'Audio Control', 'Banish', 'Biokinesis', 'Camouflage',
       'Changing Armor', 'Clairvoyance', 'Cloaking', 'Cold Resistance',
       'Cryokinesis', 'Danger Sense', 'Darkforce Manipulation',
       'Death Touch', 'Density Control', 'Dexterity',
       'Dimensional Awareness', 'Dimensional Travel', 'Duplication',
       'Durability', 'Echolocation', 'Elasticity', 'Electrical Transport',
       'Electrokinesis', 'Element Control',
       'Elemental Transmogrification', 'Empathy', 'Energy Absorption',
       'Energy Armor', 'Energy Beams', 'Energy Blasts',
       'Energy Constructs', 'Energy Manipulation', 'Energy Resistance',
       'Enhanced Hearing', 'Enhanced Memory', 'Enhanced Senses',
       'Enhanced Sight', 'Enhanced Smell', 'Enhanced Touch',
       'Fire Control', 'Fire Resis

In [8]:
for col in cols_to_make:
   df_powers[col] = df_powers['Powers'].str.contains(col)
df_powers.head()

/var/folders/yt/hk1fgcjn4w16b997bctnz_kr0000gn/T/ipykernel_59035/3448704673.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_powers[col] = df_powers['Powers'].str.contains(col)
/var/folders/yt/hk1fgcjn4w16b997bctnz_kr0000gn/T/ipykernel_59035/3448704673.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_powers[col] = df_powers['Powers'].str.contains(col)
/var/folders/yt/hk1fgcjn4w16b997bctnz_kr0000gn/T/ipykernel_59035/3448704673.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of cal

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

   Accelerated Healing  Adaptation  Agility  Animal Attributes  \
0                False       False     True              False   
1                 True       False    False              False   
2                 True       False     True              False   
3                False       False    False              False   
4                 True       False    False              False   

   Animal Control  Animal Oriented Powers  Animation  Anti-Gravity  ...  \
0           False                   False      False         False  ...   
1           False                   False      False         False  ...   
2           False                   False      False         False  ...   
3           False                   False      False         False  ...   
4           False                   False       True         False  ...   

   Vision - Thermal  Vision - X-Ray  Vitakinesis  Wallcrawling  Water Control  \
0             False           False        False         False          False   
1             False           False        False         False          False   
2             False           False        False         False          False   
3             False           False        False         False          False   
4             False           False        False         False          False   

   Weapon-based Powers  Weapons Master  Weather Control  Web Creation  \
0                False           False            False         False   
1                False           False            False         False   
2                False            True            False         False   
3                False           False            False         False   
4                False           False            False         False   

   Wind Control  
0         False  
1         False  
2         False  
3         False  
4         False  

[5 rows x 169 columns]

### Import and clean info csv

In [9]:
df_info = pd.read_csv('Data/superhero_info - superhero_info.csv')
df_info.info()
df_info.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [10]:
df_info[['hero_names','publisher']] = df_info['Hero|Publisher'].str.split('|', expand=True)
df_info = df_info.drop(columns = 'Hero|Publisher')
df_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements     hero_names  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [11]:
measurements = df_info.loc[0,"Measurements"]
print(type(measurements))
measurements

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [12]:
measurements = measurements.replace("'",'"')
measurements

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [13]:
measurements_fixed = json.loads(measurements)
print(type(measurements_fixed))
measurements_fixed

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [14]:
height_weight = df_info['Measurements'].apply(pd.Series)
height_weight

0
0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
..                                            ...
458   {'Height': '183.0 cm', 'Weight': '83.0 kg'}
459   {'Height': '165.0 cm', 'Weight': '52.0 kg'}
460    {'Height': '66.0 cm', 'Weight': '17.0 kg'}
461   {'Height': '170.0 cm', 'Weight': '57.0 kg'}
462   {'Height': '185.0 cm', 'Weight': '81.0 kg'}

[463 rows x 1 columns]

In [15]:
## use .str.replace to replace all single quotes
df_info['Measurements'] = df_info['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
df_info['Measurements'] = df_info['Measurements'].apply(json.loads)
df_info['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [16]:
height_weight = df_info['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [17]:
df_info_hw = pd.concat((df_info, height_weight), axis = 1)
df_info_hw.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements  hero_names  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien   

           publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg

In [18]:
df_info_hw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Gender        463 non-null    object
 1   Race          463 non-null    object
 2   Alignment     463 non-null    object
 3   Hair color    463 non-null    object
 4   Eye color     463 non-null    object
 5   Skin color    463 non-null    object
 6   Measurements  463 non-null    object
 7   hero_names    463 non-null    object
 8   publisher     463 non-null    object
 9   Height        463 non-null    object
 10  Weight        463 non-null    object
dtypes: object(11)
memory usage: 39.9+ KB


In [19]:
## save the 2 new columns into the dataframe
df_info_hw[['Height (cm)','cm']] = df_info_hw['Height'].str.split(' ',expand=True)
df_info_hw[['Weight (kg)','kg']] = df_info_hw['Weight'].str.split(' ',expand=True)
df_info_hw.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements  hero_names  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien   

           publisher    Height    Weight Height (cm)  cm Weight (kg)  kg  
0      Marvel Comics  203.0 cm  441.0 kg       203.0  cm       441.0  kg  
1  Dark Horse Comics  191.0 cm   65.0 kg       191.0  cm        65.0  kg

In [20]:
## drop the original column 
df_info_hw = df_info_hw.drop(columns=['Height', 'Weight', 'Measurements','cm', 'kg'])
df_info_hw.head(2)

Gender           Race Alignment Hair color Eye color Skin color  hero_names  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           publisher Height (cm) Weight (kg)  
0      Marvel Comics       203.0       441.0  
1  Dark Horse Comics       191.0        65.0

In [21]:
df_info_hw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Gender       463 non-null    object
 1   Race         463 non-null    object
 2   Alignment    463 non-null    object
 3   Hair color   463 non-null    object
 4   Eye color    463 non-null    object
 5   Skin color   463 non-null    object
 6   hero_names   463 non-null    object
 7   publisher    463 non-null    object
 8   Height (cm)  463 non-null    object
 9   Weight (kg)  463 non-null    object
dtypes: object(10)
memory usage: 36.3+ KB


In [22]:
df_info_hw['Height (cm)'] = df_info_hw['Height (cm)'].astype(float)
df_info_hw['Weight (kg)'] = df_info_hw['Weight (kg)'].astype(float)
df_info_hw.head(2)

Gender           Race Alignment Hair color Eye color Skin color  hero_names  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           publisher  Height (cm)  Weight (kg)  
0      Marvel Comics        203.0        441.0  
1  Dark Horse Comics        191.0         65.0

In [23]:
df_info_power = pd.merge(df_powers, df_info_hw, how='inner', on = 'hero_names')
df_info_power.head()

hero_names                                             Powers  \
0         A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2       Abin Sur                                 Lantern Power Ring   
3    Abomination  Accelerated Healing,Intelligence,Super Strengt...   
4  Absorbing Man  Cold Resistance,Durability,Energy Absorption,S...   

   Accelerated Healing  Adaptation  Agility  Animal Attributes  \
0                 True       False    False              False   
1                 True       False     True              False   
2                False       False    False              False   
3                 True       False    False              False   
4                False       False    False              False   

   Animal Control  Animal Oriented Powers  Animation  Anti-Gravity  ...  \
0           False                   False      False         False  ...   
1           False                   False      False         False  ...   
2           False                   False      False         False  ...   
3           False                   False       True         False  ...   
4           False                   False      False         False  ...   

   Wind Control  Gender               Race  Alignment  Hair color  Eye color  \
0         False    Male              Human       good     No Hair     yellow   
1         False    Male      Icthyo Sapien       good     No Hair       blue   
2         False    Male            Ungaran       good     No Hair       blue   
3         False    Male  Human / Radiation        bad     No Hair      green   
4         False    Male              Human        bad     No Hair       blue   

   Skin color          publisher  Height (cm)  Weight (kg)  
0     Unknown      Marvel Comics        203.0        441.0  
1        blue  Dark Horse Comics        191.0         65.0  
2         red          DC Comics        185.0         90.0  
3     Unknown      Marvel Comics        203.0        441.0  
4     Unknown      Marvel Comics        193.0        122.0  

[5 rows x 178 columns]

## II. Use your combined DataFrame to answer the following questions.

1. Compare the average weight of super powers who have Super Speed to those who do not.
2. What is the average height of heroes for each publisher?


In [24]:
speed_filter = df_info_power['Super Speed'] == True

In [25]:
df_info_power[speed_filter]['Weight (kg)'].mean()

129.40404040404042

In [26]:
df_info_power[~speed_filter]['Weight (kg)'].mean()

101.77358490566037

#### The average weight of superheros with super speed is 129 kg versus those without which is 102 kg.

#### The average height of superheros by publisher:

In [27]:
df_info_power.groupby(by = 'publisher')['Height (cm)'].mean()

publisher
DC Comics            181.923913
Dark Horse Comics    176.909091
George Lucas         159.600000
Image Comics         211.000000
Marvel Comics        191.546128
Shueisha             171.500000
Star Trek            181.500000
Team Epic TV         180.750000
Unknown              178.000000
Name: Height (cm), dtype: float64